services/galley/src/Galley/API/MLS/Reset.hs

```diff
@@ -0,0 +1,166 @@
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Galley.API.MLS.Reset
+  ( resetMLSConversation,
+    resetLocalMLSConversation,
+    ResetConversationStaticErrors,
+  )
+where
+
+import Control.Monad.Codensity hiding (reset)
+import Data.Id
+import Data.Qualified
+import Galley.API.MLS.Enabled
+import Galley.API.MLS.SubConversation
+import Galley.API.MLS.Util
+import Galley.API.Util
+import Galley.Data.Conversation.Types
+import Galley.Effects
+import Galley.Effects.ConversationStore
+import Galley.Effects.FederatorAccess
+import Galley.Effects.MemberStore
+import Galley.Env
+import Imports
+import Polysemy
+import Polysemy.Error
+import Polysemy.Input
+import Polysemy.Resource
+import Wire.API.Conversation hiding (Member)
+import Wire.API.Conversation.Protocol
+import Wire.API.Error
+import Wire.API.Error.Galley
+import Wire.API.Federation.API
+import Wire.API.Federation.API.Galley
+import Wire.API.Federation.Error
+import Wire.API.MLS.Group.Serialisation
+import Wire.API.MLS.SubConversation
+import Wire.API.Routes.Public.Galley.MLS
+
+resetMLSConversation ::
+  ( Member (Input Env) r,
+    Member (ErrorS MLSNotEnabled) r,
+    Member (ErrorS MLSStaleMessage) r,
+    Member (ErrorS ConvAccessDenied) r,
+    Member (ErrorS ConvNotFound) r,
+    Member (Error MLSProtocolError) r,
+    Member (ErrorS InvalidOperation) r,
+    Member (ErrorS MLSFederatedResetNotSupported) r,
+    Member Resource r,
+    Member ConversationStore r,
+    Member FederatorAccess r,
+    Member (Error FederationError) r,
+    Member MemberStore r,
+    Member SubConversationStore r
+  ) =>
+  Local UserId ->
+  MLSReset ->
+  Sem r ()
+resetMLSConversation lusr reset = do
+  assertMLSEnabled
+  (ctype, qcnvOrSub) <- getConvFromGroupId reset.groupId
+  foldQualified
+    lusr
+    (\lcnvOrSub -> resetLocalMLSConversation (tUntagged lusr) ctype lcnvOrSub reset)
+    (\r -> resetRemoteMLSConversation r lusr reset)
+    qcnvOrSub
+
+resetLocalMLSConversation ::
+  ( Member (Input Env) r,
+    Member (ErrorS MLSNotEnabled) r,
+    Member (ErrorS MLSStaleMessage) r,
+    Member (ErrorS ConvAccessDenied) r,
+    Member (ErrorS ConvNotFound) r,
+    Member (ErrorS InvalidOperation) r,
+    Member Resource r,
+    Member ConversationStore r,
+    Member MemberStore r,
+    Member SubConversationStore r
+  ) =>
+  Qualified UserId ->
+  ConvType ->
+  Local ConvOrSubConvId ->
+  MLSReset ->
+  Sem r ()
+resetLocalMLSConversation qusr ctype lcnvOrSub reset =
+  case tUnqualified lcnvOrSub of
+    SubConv c s -> deleteLocalSubConversation qusr (c <$ lcnvOrSub) s reset -- TODO: extract
```

Comment:
`TODO` or `FUTUREWORK`?

Reply:
Ah, thanks for catching this. I wanted to refactor this slightly to avoid invoking the API handler in this case, but I'll move it to a follow-up PR.